### PR TITLE
pkg/logger: add basic zap test and prod logger

### PR DESCRIPTION
**Pull request by jmank88**
https://app.shortcut.com/chainlinklabs/story/40148/chainlink-relay-create-logger-impl-for-testing-and-scripts

### CODE REVIEW

**Review by topliceanu — APPROVED** [review]

File: pkg/logger/logger.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	Panicw(msg string, keysAndValues ...interface{})
+	Fatalw(msg string, keysAndValues ...interface{})
+
+	Sync() error
+}
+
+// New returns a new Logger.
+func New() (Logger, error) {
+	var lvl zapcore.Level
+	err := lvl.Set(os.Getenv("LOG_LEVEL"))
```
Comment (topliceanu) [inline pkg/logger/logger.go:45]: I would prefer if log level was passed as a parameter to the constructor. 
I don't like env vars peppered around the code and the dependencies!
_my 2c_

File: pkg/logger/logger.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	Fatalf(format string, values ...interface{})
+
+	Debugw(msg string, keysAndValues ...interface{})
+	Infow(msg string, keysAndValues ...interface{})
+	Warnw(msg string, keysAndValues ...interface{})
+	Errorw(msg string, keysAndValues ...interface{})
+	Panicw(msg string, keysAndValues ...interface{})
+	Fatalw(msg string, keysAndValues ...interface{})
+
+	Sync() error
```
Comment (topliceanu) [inline pkg/logger/logger.go:39]: Can you add some documentation about when to use `Sync()` and what it does?

File: pkg/logger/logger.go (hunk 1 above)
Comment (jmank88) [reply pkg/logger/logger.go:45]: Will do! Almost did it before but was trying to capture the API exactly at first.

File: pkg/logger/logger.go (hunk 2 above)
Comment (jmank88) [reply pkg/logger/logger.go:38]: Is pointing to core `logger.Logger` and `zap.SugaredLogger` from the type level comment sufficient?

File: pkg/logger/logger.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"go.uber.org/zap/zaptest"
+	"go.uber.org/zap/zaptest/observer"
+)
+
+// Logger is a minimal subset of smartcontractkit/chainlink/core/logger.Logger implemented by go.uber.org/zap.SugaredLogger
+type Logger interface {
+	With(args ...interface{}) Logger
```
Comment (jmank88) [inline pkg/logger/logger.go:14]: This self reference is a problem. Although core/logger.Logger also contains a `With(args ...interface) Logger`, the method signatures don't match. I don't think there is any way to express this with generics, but a transitional helper func might be possible, like `logger.With(Logger, args ...interface{})` :thinking:

Comment (jmank88) [reply pkg/logger/logger.go]: Added `func With`, which is not pretty, but fixes the problem. Open to better ideas :shrug:

File: pkg/logger/logger.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	*zap.SugaredLogger
+}
+
+func (l *logger) with(args ...interface{}) Logger {
+	return &logger{l.SugaredLogger.With(args...)}
+}
+
+var typeOfLogger = reflect.TypeOf((Logger)(nil))
+
+func With(l Logger, args ...interface{}) Logger {
```
Comment (jmank88) [inline pkg/logger/logger.go:88]: - [ ] doc
- [ ] test

...or replace with some alternative?

File: pkg/logger/logger.go (hunk 2 above)
Comment (topliceanu) [reply pkg/logger/logger.go:36]: Yeah!

File: pkg/logger/logger.go (hunk 3 above)
Comment (topliceanu) [reply pkg/logger/logger.go]: I know this pain. I've had to do this https://github.com/smartcontractkit/chainlink-terra/blob/develop/cmd/monitoring/main.go#L74-L82 
I think a separate function `With(args ...interface) Logger` is fine!

File: pkg/logger/logger.go (hunk 4 above)
Comment (topliceanu) [reply pkg/logger/logger.go:88]: I think this works. I pick doc, because YOLO!

**Review by topliceanu — APPROVED** [review]

**Comment by topliceanu** [comment]
`core/logger` is the only dependency `chainlink-relay/pkg/monitoring` has on `core`. I wanted to remove it and replace it with zap directly for months now! I appreciate you adding this!

File: pkg/logger/logger.go (hunk 4 above)
Comment (jmank88) [reply pkg/logger/logger.go]: Done, and it needed a bit of fixing too.

**Review by topliceanu — APPROVED** [review]
lgtm

**Pull request merged**